service/src/main/java/org/cbioportal/service/impl/CacheServiceImpl.java

```diff
@@ -31,4 +44,61 @@ public void clearCaches(boolean clearSpringManagedCache) {
         // version of the r-library and may result in problems in the running instance
         // when flushing the cashes, we do not handle these caches in this service.
     }
+
+    // This evicts keys from the general and static caches when updating/adding/deleting a study.
+    public void clearCachesForStudy(String studyId, List<String> allStudyIds, boolean clearSpringManagedCache) throws CacheOperationException {
+
+        // Flush Spring-managed caches (only when cache strategy has been defined).
+        if (clearSpringManagedCache) {
+            attemptEvictSpringManagedCache(buildRegex(studyId, allStudyIds));
+        }
+        
+        // Flush cache used for user permission evaluation.
+        cacheMapUtil.initializeCacheMemory();
+    }
+    
+    private void attemptEvictSpringManagedCache(String pattern) throws CacheOperationException {
+        try {
+            if (cacheManager != null) {
+                cacheManager.getCacheNames().stream()
+                    .forEach(cacheName -> {
+                        cacheUtils.evictByPattern(cacheName, pattern);
+                    });
+            }
+        } catch (RuntimeException e) {
+            e.printStackTrace();
+            LOG.error("Error while evicting cache." + e.getMessage());
+            throw new CacheOperationException("Error while evicting cache.", e);
+        }
+    }
+    
+    // Regex that selects keys that match id of deleted study
+    // or lacking any study id completely. For example:
+    // ^(?=.*study_id_1).*|^(?!.*study_id_1)(?!.*study_id_2)(?!.*study_id_3).*
+    // https://stackoverflow.com/a/8240998/11651683
+    private String buildRegex(String studyId, List<String> allStudyIds) {
+        
+        // make sure imported studyId is in all studies.
+        List<String> all = new ArrayList<>();
+        all.add(studyId);
+        all.addAll(allStudyIds);
+
+        StringBuilder stringBuilder = new StringBuilder();
+        stringBuilder.append("^");
+        stringBuilder.append("(?=.*");
+        stringBuilder.append(studyId);
+        stringBuilder.append(").*");
+
+        stringBuilder.append("|");
+        stringBuilder.append("^");
+        
+        all.forEach(id -> {
+            stringBuilder.append("(?!.*");
+            stringBuilder.append(id);
+            stringBuilder.append(")");
+        });
+        stringBuilder.append(".*");
+        return stringBuilder.toString();
```

Comment:
Not a huge fan of all this StringBuilder stuff - I don't think we need to make regex any harder to read. Here's what this looks like if you use `String.format`. You can take it or leave it.

```suggestion
        String allIdsRegex = all.stream()
            .map(id -> "(?!.*" + id + ")")
            .collect(Collectors.joining(""));
        
        return String.format("^(?=.*%s).*|^%s.*", studyId, allIdsRegex);
```

Reply:
Oooooh much nicer! I used StringBuilder since I remember building strings is more memory efficient this way. But your solution is much nicer. Thanks I will remember this.